### PR TITLE
Fix gpexpand flaky test cases

### PR DESCRIPTION
**Pull request by kainwen**
Gpexpand test cases often need to stop the cluster.
There is race condition in previous code:

1. it invokes pkill
2. it tests if cluster is there
3. if in step 2 it founds cluster is on, it invokes gpstop

The race condition is that the cluster gets down during step 2
and step 3, which leads to exception when doing gpstop while cluster
is down.

This commit fixes flaky gpexpand test cases by checking process
does not exists after killing it.

-------

Devpipeline is green for about 40 times. https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/remove_kill_gpexpand/jobs/gpexpand/builds/42

And please check that the "deep bug" is explained correctly in the comment https://github.com/greenplum-db/gpdb/pull/7245#issuecomment-477519558

As @pivotal-ning-yu points out, there are many other places using kill to stop dbs.  If that explanation is right, we might fix like this commit and to totally refactor all the places use `pkill` to 
have a global refactor design.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by kainwen** [comment]
This commit is to fix the issue #7229

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 1]
```diff
@@ -39,6 +39,57 @@
...
+                                          ctxt=REMOTE,
+                                          remoteHost=host)
+
+    def validate(self, expected_rc=0):
+        super(KillCommand, self).validate()
+        start_time = datetime.now()
+        end_time = start_time
+        killFinished = False
+        while (not killFinished and
+               ((end_time - start_time).total_seconds() < self.waitTime)):
```
Comment (danielgustafsson) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: I don't know Python, so I might be missing something obvious, but doesn't this require a small sleep to avoid busy-looping? Hammering as fast as we can on pgrep seems pretty suboptimal.

Comment (kainwen) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: > I don't know Python, so I might be missing something obvious, but doesn't this require a small sleep to avoid busy-looping? Hammering as fast as we can on pgrep seems pretty suboptimal.

Good catch. I should add some small sleep.

Thanks for the review! @danielgustafsson

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 2]
```diff
@@ -39,6 +39,57 @@
...
+    It overrides the validate method. In the
+    validate stage, it invokes pgrep command
+    to verify that the pkill command has taken
+    effect.
+    """
+    def __init__(self, procName, host, waitTime=90):
+        self.procName = procName
+        self.waitTime = waitTime
+        self.host = host
+        name = "pkill " + procName
```
[the remark this answers is not in the document]
Comment (kainwen) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: > I know this isn't the fault of this PR, but this seems pretty wrong to me. I routinely run several postgres instances on my machine that I absolutely don't want randomly killed.. Can't we kill directly on the pids in question?

@danielgustafsson Thanks for your comments.

I understand.

Usually, we seldom run this case locally, so it is acceptable just in this case?

[the remark this answers is not in the document]
Comment (danielgustafsson) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: It's not introduced in this PR so it doesn't need to be solved in this PR either. I just wanted to bring up the issue and a hope that we at some point will refactor these behave tests such that they are more friendly to a multi-user environment.

[the remark this answers is not in the document]
Comment (danielgustafsson) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: Pinging @jchampio to get this into his mailbox since he is currently looking at these things.

**Comment by jchampio** [comment]
Our thought was that the pattern
```gherkin
Given the database is killed on "<all hosts>"
  And [something else]
  And the database is not running
```
can just be replaced with
```gherkin
Given the database is not running
  And [something else]
```
It seems like the cases we've seen so far don't need the kill step to begin with, since we're going to be waiting for gpstop anyway in the `database is not running` step if the database isn't stopped by the time we get to it. Is there a reason that we need an explicit per-host kill for these tests, or can we just get rid of it?

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jchampio) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: > I just wanted to bring up the issue and a hope that we at some point will refactor these behave tests such that they are more friendly to a multi-user environment.

That's the plan, yeah. That same conversation came up in #7235.

[the remark this answers is not in the document]
Comment (danielgustafsson) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: Virtual high-fives all around for that!

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 3]
```diff
@@ -39,6 +39,58 @@
...
+
+class KillCommand(Command):
+    """
+    This class is a simple wrapper of Command.
+    It overrides the validate method. In the
+    validate stage, it invokes pgrep command
+    to verify that the pkill command has taken
+    effect.
+    """
+    def __init__(self, procName, host, waitTime=90):
```
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: The default waiting period per segment for `gpstop` (see `-t` option) is 600 seconds.  I am not sure, if one stops a 10 node cluster, whether the total wait time is 600 seconds or 600*10 seconds.  But, in any case, perhaps the longer timeout is warranted.  I am not sure how long is "too long" for this, but why did you choose 90 seconds?

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dkrieger-pvtl) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: This would be addressed by using `gpstop` instead of `pkill postgres`, which is discussed as a Conversation on this PR.

**Comment by dkrieger-pvtl** [comment]
> It seems like the cases we've seen so far don't need the kill step to begin with, since we're going to be waiting for gpstop anyway in the `database is not running` step if the database isn't stopped by the time we get to it. Is there a reason that we need an explicit per-host kill for these tests, or can we just get rid of it?

Yes, I assume that if `pkill [TERM]` works, then `gpstop` would as well.  That is safer as well, as it does not interfere with other `postgres` processes running on the machine.  As @jchampio points out, why not just use `gpstop -M fast -a `, as is currently done in the step `database is not running`?

Your implementation itself appears correct as it is, but I think takes the wrong approach.  There might be a use case for `pkill postgres`(for instance, to summarily eliminate a host from a cluster for data recovery testing), but the current uses of that step, in both `gpmovemirrors` and `gpexpand` appear to not need it.  So we could even remove that step from our codebase.

**Comment by dkrieger-pvtl** [comment]
I believe we will want to backport this to 6X_STABLE as well.

**Comment by dkrieger-pvtl** [comment]
Note the gpexpand.feature test in 6X has been failing on the Scenario `Verify policy of writable external table is correctly updated after redistribution` four times in a row, so you could potentially use that as a red-green test case for your changes.

**Comment by pivotal-ning-yu** [comment]
Currently we are doing `pkill` then `gpstop`, I'm thinking whether it's better to change it to `gpstop` then `pkill`.  The point for `pkill` is that it can be used to cleanup the processes of new segments which failed to add to the cluster in previous tests.

**Comment by kainwen** [comment]
@jchampio @dkrieger-pvtl 
Thanks for the review.

I tried another version that simply removes the kill instruction.

It seems that if the kill command is to send to multi hosts. It cannot be replaced by single gpstop if there are some other remaining postgres process on hosts.

In the dev pipeline, it fails https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/remove_kill_gpexpand/jobs/gpexpand/builds/1

I have not looked into it deeply.  Any thoughts?

**Comment by dkrieger-pvtl** [comment]
>  The point for `pkill` is that it can be used to cleanup the processes of new segments which failed to add to the cluster in previous tests.

This seems like a valid case for pkill, but a more specific search string other than `pkill postgres` such as `pkill postgres | grep MY_DATA_DIRECTORY` would probably be better.  But I wonder if there's another way to handle this, such as tracking specifically which postgres processes are created and killing just those.  

Furthermore, skimming our Behave tests, this doesn't seem like the common case...and that most often, we should be able to use `gpstop`.  I understand the need to want a clean slate for each scenario but the common case should be all passing...

**Comment by dkrieger-pvtl** [comment]
> I tried another version that simply removes the kill instruction.
> 
> It seems that if the kill command is to send to multi hosts. It cannot be replaced by single gpstop if there are some other remaining postgres process on hosts.
> 
> In the dev pipeline, it fails https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/remove_kill_gpexpand/jobs/gpexpand/builds/1
> 
> I have not looked into it deeply. Any thoughts?

The gpexpand tests run on a ccp cluster, and if you look at the `gpexpand` job in our pipeline, it is running the Behave tests on the mdw, where the master should be.  So I'm not sure why the step `And the database is not running` is not bringing down the cluster.  It's clearly not if you look at the failing tests.

But I think the path forward would be to figure out why that step isn't doing so...the tests will be much cleaner.

Does your current implementation(`pkill postgres` with a synchronous timeout to wait) reliably allow all tests to pass?  Another approach, since the current gpexpand feature tests fail pretty often, would be to submit this fix and then update it later to clean up this step.  Your original implementation solves the problem at hand...

**Comment by jchampio** [comment]
> But I think the path forward would be to figure out why that step isn't doing so...the tests will be much cleaner.

Yeah, that seems like a pretty important bug to fix, and `pkill`ing everything might be papering over it.

**Comment by jchampio** [comment]
> Another approach, since the current gpexpand feature tests fail pretty often, would be to submit this fix and then update it later to clean up this step. 

I'm less comfortable with this; once the tests go green it's generally out of mind. Now that we're suspicious of a deeper bug, we should fix it.

**Comment by dkrieger-pvtl** [comment]
> > Another approach, since the current gpexpand feature tests fail pretty often, would be to submit this fix and then update it later to clean up this step.
> 
> I'm less comfortable with this; once the tests go green it's generally out of mind. Now that we're suspicious of a deeper bug, we should fix it.

I agree now, thinking about it.  Most of the tests should be able to use `gpstop` and we should figure out why the current tests fail with it.

**Comment by kainwen** [comment]
@dkrieger-pvtl @jchampio @danielgustafsson 

I think I have found the "deep bug".  We suspect that it is the race condition between kill and gptsop (please see the issue https://github.com/greenplum-db/gpdb/issues/7229 ). 

Let me answer two questions:

Q1: Why other cases do not show such race condition for so long a time?
Q2: Why always the case of writable external table flaky?

A1:
The impl python code of "And the database is not running" is 
```python
@given('the database is not running')
@when('the database is not running')
def impl(context):
    stop_database_if_started(context)
    if has_exception(context):
        raise context.exception
```
It will invoke the function `stop_database_if_started` and here it only provides `context`, all other parameters will use default values.

Let's see `stop_database_if_started`'s code:

```python
def stop_database_if_started(context):
    if check_database_is_running(context):
        stop_database(context)
```

It first invokes `check_database_is_running` and based on the return value then invoke `gpstop` or not. Also, note here invoke `check_database_is_running` only pass context, so all other parameters will use default values.

`check_database_is_running` invokes `chk_local_db_running` by PGPORT and MASTER_DIR locally.

And only writable external table case and the case before it use "localhost".
Other cases all use mdw, sdw1 which are vms on gcp/aws. So in the docker, which test is running, this stop_database_if_started will always detect not-running, so no gpstop invoked.

------------------

A2:

writable_External_table is right after the previous case uses "localhost", and enlarge it into 6 segments.
So I guess: kill postgres will take long time and gpstop will invoke, so race condition.

------------------

Summary: the PR that check kill is OK to fix it. And I run about 20 times, the pipeline is always green.

https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/fix_flaky_gpexpand/jobs/gpexpand/builds/19

**Comment by dkrieger-pvtl** [comment]
> So I guess: kill postgres will take long time and gpstop will invoke, so race condition.
> 
> Summary: the PR that check kill is OK to fix it. And I run about 20 times, the pipeline is always green.
> 
> https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/fix_flaky_gpexpand/jobs/gpexpand/builds/19

As discussed above, the suggested fix indeed works because it essentially makes the asynchronous `pkill(postgres)` synchronous by looping until it has taken effect.  It is a correct implementation but has the following drawbacks:

1.  as @danielgustafsson points out, you unnecessarily also bring down any other `postgres` instances on the same machine.

1. it is essentially reimplementing the functionality of `gpstop` which ultimately will even send a `kill -9` to all the processes, ensuring that the cluster is indeed stopped.  There is no need to reimplement this functionality.

As I mentioned above, the only valid use case for using `pkill` directly is if a test somehow creates "orphaned" postgres instances that aren't part of our cluster.  The comments mention one such case. But all other cases can and should be reimplemented using `gpstop`.  Furthermore, in such cases, the `pkill` should be more specific and only kill the test's instances, not all those on the machine.

As far as the discussion on why the race occurs only in some circumstances and others, that is the nature of race conditions and timing differences indeed often expose them.  

Can you please try eliminating `pkill` in favor of `gpstop` as an alternate, cleaner solution?

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 4]
```diff
@@ -39,6 +39,58 @@
...
+        self.waitTime = waitTime
+        self.host = host
+        name = "pkill " + procName
+        cmdStr = name + " || true"
+        super(KillCommand, self).__init__(name=name,
+                                          cmdStr=cmdStr,
+                                          ctxt=REMOTE,
+                                          remoteHost=host)
+
+    def validate(self, expected_rc=0):
```
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: There is a bug here too...only processes that match the name that existed at the **start** of the command are checked...if a process that matches the name is created AFTER the `pkill` but before the `pgrep` runs and stays alive for longer than the timeout, the `validate` will fail.  This race is currently 5 seconds long...if the `pkill` isn't immediate, there are 5 seconds for another so-named process to run.

So this solution introduces race conditions of its own.

**Comment by pivotal-ning-yu** [comment]
FYI, a similar issue happened on gpaddmirrmors:
https://prod.ci.gpdb.pivotal.io/teams/main/pipelines/5X_STABLE/jobs/MM_gpaddmirrors/builds/195

```
Scenario: gpaddmirrors when the primaries have data 
    Given a working directory of the test as '/tmp/gpaddmirrors'                             # 0.000s
    And the user runs command "rm -rf /tmp/gpaddmirrors/*"                                   # 0.054s
    And the database is killed on hosts "mdw,sdw1"                                           # 1.305s
    And a cluster is created with no mirrors on "mdw" and "sdw1"                             # 12.090s
      Traceback (most recent call last):
        File "build/bdist.linux-x86_64/egg/behave/model.py", line 1173, in run
          match.run(runner.context)
        File "build/bdist.linux-x86_64/egg/behave/model.py", line 1589, in run
          self.func(context, *args, **kwargs)
        File "/home/gpadmin/gpdb_src/gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py", line 5592, in impl
          _create_cluster(context, master_host, segment_host_list)
        File "/home/gpadmin/gpdb_src/gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py", line 5587, in _create_cluster
          testcluster.create_cluster(with_mirrors=False)
        File "/home/gpadmin/gpdb_src/gpMgmt/test/behave_utils/cluster_setup.py", line 160, in create_cluster
          raise Exception("Failed initializing the cluster. Look into gpAdminLogs for more information")
      Exception: Failed initializing the cluster. Look into gpAdminLogs for more information

...

20190328:18:09:45:gpinitsystem:mdw:gpadmin-[FATAL]:-Host sdw1 has an active database process on port = 20501 Script Exiting!
```

I hijacked on the job and believe that it shares the same race condition issue with current one:
1. a previous segment was running on port 20501;
2. pkill them;
3. create a new segment on port 20501, but previous segment hasn't stopped so the port is not released yet;

I'm afraid other utilities have similar issues in their behave tests.  I would suggest a full scan on all the behave tests and have a overall fix instead just focusing on gpexpand.

File: gpMgmt/test/behave/mgmt_utils/gpexpand.feature [hunk 5]
```diff
@@ -194,9 +193,8 @@ Feature: expand the cluster by adding more segments
     @gpexpand_verify_writable_external_redistribution
     Scenario: Verify policy of writable external table is correctly updated after redistribution 
         Given a working directory of the test as '/data/gpdata/gpexpand'
-        And the database is killed on hosts "localhost"
-        And a temporary directory under "/data/gpdata/gpexpand/expandedData" to expand into
         And the database is not running
+        And a temporary directory under "/data/gpdata/gpexpand/expandedData" to expand into
         And the cluster is generated with "1" primaries only
         And database "gptest" exists
         And the user create a writable external table with name "ext_test"
```
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/gpexpand.feature]: Thanks; this is the code change that should work and remove the (unnecessary) step that introduces the race.  Please update all the relevant gpexpand test cases; there are many more.  Hopefully, this idiom will work on all the gpexpand tests.

Comment (kainwen) [reply gpMgmt/test/behave/mgmt_utils/gpexpand.feature]: > Thanks; this is the code change that should work and remove the (unnecessary) step that introduces the race. Please update all the relevant gpexpand test cases; there are many more. Hopefully, this idiom will work on all the gpexpand tests.

I do not think so.

Have your read my comment https://github.com/greenplum-db/gpdb/pull/7245#issuecomment-477519558 and the PR message?

For any other cases that do not build cluster on localhost, `gpstop` is not invoked.

And also, as @pivotal-ning-yu points out, there are many other cases use kill.

So I think we should do a total refactor of kill for all scripts in future. Not for this PR.

**Comment by yydzero** [comment]
@dkrieger-pvtl @jchampio I just talked with @kainwen to understand more context about this PR, as a couple of teammates from other teams have been blocked by this flaky test to merge to master. (We are experimenting guideline that no merge on RED pipeline).

My understanding is that we all agree the RCA and  this PR will fix the flaky test case in question, and also reached agreement on the way to fix this particular issue.  So I suggest we merge this PR sooner than later to fix the most fragile flaky test to make pipeline more stable and unblock others to merge to master.

Next question is whether we leverage same way to enhance other gpexpand test cases (that is to avoid to use pkill/kill altogether). This is a separated question than the issue in this PR, as test cases involved in this PR is running on local host, while other test cases are using remote cluster on gcp. And other test cases seems happy so far (maybe because it is remote, so pkill has much high chance to finish its job before next statement runs) and we have not find flaky test cases so far.

While we also explored the code a bit to try to understand why it is introduced in the first place. Our guess is that pkill is introduced to cleanup broken cluster environment which is caused by previous test cases. Without pkill in such situation, one broken (not flaky test, it caught a bug) test will result in many test failures (This also impact the fix in this PR). Feel this makes sense. While I am also fine if we finally choose the path to remove all pkill, and endure that one failed case results in many test failures due to unclean environment caused by previously failed case.  While this needs broader discussion and will take longer time.

So in summary, root cause is found, and let us merge this PR to fix the hottest issue to unblock others to merge. And then we have a discussion on the other question list above. I will try to setup a call for it later.

Thoughts?

**Comment by dkrieger-pvtl** [comment]
Other than the race conditions on this current implementation(see comments to code above  https://github.com/greenplum-db/gpdb/pull/7245/files#r270122137), it does workaround the current failures.  But it doesn't resolve them in the sense that it uses `SIGTERM` to stop the processes and not `SIGKILL`(which has its own issues as @danielgustafsson points out: you'll stop ALL `postgres` on the machine).

I think a better solution iis to make sure `gpstop` is being called at the beginning of each test; it need not use the current Behave step.  I thought that the ` the database is not running` actually runs on ccp clusters on `mdw`, though I am apparently missing something as this does not work.

If exceptions need to be made, they should be explicit on particular tests that need them, not blanket methods to ensure tests are independent.  That's what the test setup is for...in fact, you should use the `before_scenario/after_scenario` (see gpMgmt/test/behave/mgmt_utils/environment.py).

**Comment by kainwen** [comment]
@dkrieger-pvtl @jchampio @yydzero @pivotal-ning-yu 
Sorry. The "deep bug" analysis from me is wrong. I make a huge mistake.

I will spend more time to find why always such case.
